src/Workspaces/Core/Portable/VirtualChars/VirtualChar.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.VirtualChars
+{
+    /// <summary>
+    /// The Regex parser wants to work over an array of characters, however this array of characters
+    /// is not the same as the array of characters a user types into a string in C# or VB.  For example
+    /// In C# someone may write: @"\z".  This should appear to the user the same as if they wrote "\\z"
+    /// and the same as "\\\u007a".  However, as these all have wildly different presentations for the
+    /// user, there needs to be a way to map back the characters it sees ( '\' and 'z' ) back to the 
+    /// ranges of characters the user wrote.  
+    /// 
+    /// VirtualChar serves this purpose.  It contains the interpretted value of any language character/
+    /// character-escape-sequence, as well as the original SourceText span where that interpretted 
+    /// character was created from.  This allows the regex engine to both process regexes from any
+    /// language uniformly, but then also produce trees and diagnostics that map back properly to
+    /// the original source text locations that make sense to the user.
+    /// </summary>
+    internal struct VirtualChar : IEquatable<VirtualChar>
+    {
+        public readonly char Char;
+        public readonly TextSpan Span;
+
+        public VirtualChar(char @char, TextSpan span)
+        {
+            if (span.IsEmpty)
+            {
+                throw new ArgumentException();
```

Comment:
Use `ArguementOutOfRangeException( $"{ nameof( span ) } was expected to be non empty.")` or at least add which parameter it is.

Reply:
Sure.